[C, C++ & ObjC codegen] Incompatible pointer types used in codegen (an error on GCC 14)

### Description

I tried to build [Nish](https://github.com/thindil/nish) on commit [3087b0b41c7cc582592c8cfa6728857532862160](https://github.com/thindil/nish/commit/3087b0b41c7cc582592c8cfa6728857532862160). GCC spat out `-Wincompatible-pointer-types` errors. This used to be a warning, but was [promoted to an error in GCC 14](https://gcc.gnu.org/gcc-14/porting_to.html).

### Nim Version

Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2024-05-31
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: c1f31cedbb81c4ba6d82d85d6499dd812e7c6ab8
active boot switches: -d:release

### Current Output

```text
$ nimble build -d:release
  Verifying dependencies for nish@0.8.0
     Info:  Dependency on contracts@>= 0.2.2 already satisfied
  Verifying dependencies for contracts@0.2.2
 Installing nimassets@>= 0.2.4
Downloading https://github.com/xmonader/nimassets using git
   Warning: The package has no tagged releases, downloading HEAD instead.
  Verifying dependencies for nimassets@0.2.4
     Info:  Dependency on zstd@>= 0.5.0 already satisfied
  Verifying dependencies for zstd@0.9.0
 Installing nimassets@0.2.4
   Building nimassets/nimassets using c backend
/tmp/nimble_34728/githubcom_xmonadernimassets_0.2.4/src/nimassets.nim(3, 8) Warning: use the nimble packages `malebolgia`, `taskpools` or `weave` instead; threadpool is deprecated [Deprecated]
/home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c: In function ‘handleFileWrapper__nimassets_u849’:
/home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c:1185:26: error: passing argument 1 of ‘nimFlowVarSignal’ from incompatible pointer type [-Wincompatible-pointer-types]
 1185 |         nimFlowVarSignal(colontmp__2);
      |                          ^~~~~~~~~~~
      |                          |
      |                          tyObject_FlowVarObj__VSrjOaNbUpyhC9bP2xuCr9cw *
/home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c:225:99: note: expected ‘tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA *’ but argument is of type ‘tyObject_FlowVarObj__VSrjOaNbUpyhC9bP2xuCr9cw *’
  225 | N_LIB_PRIVATE N_NIMCALL(void, nimFlowVarSignal)(tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA* fv_p0);
      |                                                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
/home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c: In function ‘generateDirAssetsSpawn__nimassets_u299’:
/home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c:1513:130: error: passing argument 1 of ‘nimFlowVarCreateSemaphore’ from incompatible pointer type [-Wincompatible-pointer-types]
 1513 |                                                                                                 nimFlowVarCreateSemaphore(scratch.fv1);
      |                                                                                                                           ~~~~~~~^~~~
      |                                                                                                                                  |
      |                                                                                                                                  tyObject_FlowVarObj__VSrjOaNbUpyhC9bP2xuCr9cw *
/home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c:204:108: note: expected ‘tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA *’ but argument is of type ‘tyObject_FlowVarObj__VSrjOaNbUpyhC9bP2xuCr9cw *’
  204 | N_LIB_PRIVATE N_NIMCALL(void, nimFlowVarCreateSemaphore)(tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA* fv_p0);
      |                                                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread -I/home/termer/.nimble/pkgs2/zstd-0.9.0-20b23158e94f01ea0c4bf419a21b0feabe70bf31/zstd/deps/zstd -O3 -fno-strict-aliasing -fno-ident -fno-math-errno   -I'/home/termer/.choosenim/toolchains/nim-#devel/lib' -I/tmp/nimble_34728/githubcom_xmonadernimassets_0.2.4/src -o /home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c.o /home/termer/.cache/nim/nimassets_r/@mnimassets.nim.c' failed with exit code: 1


       Tip: 10 messages have been suppressed, use --verbose to show them.
nimble.nim(304)          buildFromDir

    Error:  Build failed for the package: nimassets
```


### Expected Output

_No response_

### Possible Solution

The pointer can be cast to the desired pointer type, or just `void*`. With that said, it looks like GCC may be catching a legitimate case of bad codegen, because the pointer types are different.

When looking at the C structs, I see the following:

```c
struct tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA {
  RootObj Sup;
    NIM_BOOL ready;
    NIM_BOOL usesSemaphore;
    NIM_BOOL awaited;
    tyObject_Semaphore__CkiZK7FSJX59c3CxXmi1yUA cv;
    tyObject_AwaitInfo__MP6pfMizkdD8cpJBQdARdw* ai;
    NI idx;
    void* data;
    void* owner;
};
```

and

```c
struct tyObject_FlowVarObj__VSrjOaNbUpyhC9bP2xuCr9cw {
  tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA Sup;
    NimStringV2 blob;
};
```

I'm not familiar with the C representation of Nim objects, but it looks like `tyObject_FlowVarObj__VSrjOaNbUpyhC9bP2xuCr9cw` is a subtype of `tyObject_FlowVarBaseObj__9aHwT1H8xoRHbenZXyn8OYA`. In Nim land, using a subtype in place of a supertype isn't a problem of course, but it looks like they have different memory layouts in C land. It may not be a problem though, I don't know how Nim codegen handles these things.

Either way, casting to the desired type would get rid of this compiler error.

### Additional Information

GCC:

```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/x86_64-pc-linux-gnu/14.1.1/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --enable-languages=ada,c,c++,d,fortran,go,lto,m2,objc,obj-c++,rust --enable-bootstrap --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://gitlab.archlinux.org/archlinux/packaging/packages/gcc/-/issues --with-build-config=bootstrap-lto --with-linker-hash-style=gnu --with-system-zlib --enable-__cxa_atexit --enable-cet=auto --enable-checking=release --enable-clocale=gnu --enable-default-pie --enable-default-ssp --enable-gnu-indirect-function --enable-gnu-unique-object --enable-libstdcxx-backtrace --enable-link-serialization=1 --enable-linker-build-id --enable-lto --enable-multilib --enable-plugin --enable-shared --enable-threads=posix --disable-libssp --disable-libstdcxx-pch --disable-werror
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 14.1.1 20240507 (GCC)
```

---

OS: `Linux 6.9.1-arch1-1 x86_64`

--- a/compiler/spawn.nim
+++ b/compiler/spawn.nim
@@ -110,11 +110,14 @@ stmtList:
 """
 
 proc castToVoidPointer(g: ModuleGraph, n: PNode, fvField: PNode): PNode =
-  let ptrType = getSysType(g, n.info, tyPointer)
-  result = newNodeI(nkCast, fvField.info)
-  result.add newNodeI(nkEmpty, fvField.info)
-  result.add fvField
-  result.typ = ptrType
+  if g.config.backend == backendCpp:
+    result = fvField
+  else:
+    let ptrType = getSysType(g, n.info, tyPointer)
+    result = newNodeI(nkCast, fvField.info)
+    result.add newNodeI(nkEmpty, fvField.info)
+    result.add fvField
+    result.typ = ptrType
 
 proc createWrapperProc(g: ModuleGraph; f: PNode; threadParam, argsParam: PSym;
                        varSection, varInit, call, barrier, fv: PNode;

--- a/compiler/spawn.nim
+++ b/compiler/spawn.nim
@@ -109,6 +109,13 @@ stmtList:
 
 """
 
+proc castToVoidPointer(g: ModuleGraph, n: PNode, fvField: PNode): PNode =
+  let ptrType = getSysType(g, n.info, tyPointer)
+  result = newNodeI(nkCast, fvField.info)
+  result.add newNodeI(nkEmpty, fvField.info)
+  result.add fvField
+  result.typ = ptrType
+
 proc createWrapperProc(g: ModuleGraph; f: PNode; threadParam, argsParam: PSym;
                        varSection, varInit, call, barrier, fv: PNode;
                        idgen: IdGenerator;
@@ -156,8 +163,9 @@ proc createWrapperProc(g: ModuleGraph; f: PNode; threadParam, argsParam: PSym;
     if barrier == nil:
       # by now 'fv' is shared and thus might have beeen overwritten! we need
       # to use the thread-local view instead:
+      let castExpr = castToVoidPointer(g, f, threadLocalProm.newSymNode)
       body.add callCodegenProc(g, "nimFlowVarSignal", threadLocalProm.info,
-        threadLocalProm.newSymNode)
+        castExpr)
   else:
     body.add call
   if barrier != nil:
@@ -413,7 +421,8 @@ proc wrapProcForSpawn*(g: ModuleGraph; idgen: IdGenerator; owner: PSym; spawnExp
     # create flowVar:
     result.add newFastAsgnStmt(fvField, callProc(spawnExpr[^1]))
     if barrier == nil:
-      result.add callCodegenProc(g, "nimFlowVarCreateSemaphore", fvField.info, fvField)
+      let castExpr = castToVoidPointer(g, n, fvField)
+      result.add callCodegenProc(g, "nimFlowVarCreateSemaphore", fvField.info, castExpr)
 
   elif spawnKind == srByVar:
     var field = newSym(skField, getIdent(g.cache, "fv"), idgen, owner, n.info, g.config.options)

--- a/tests/c/temit.nim
+++ b/tests/c/temit.nim
@@ -4,6 +4,7 @@ discard """
 # Test the new ``emit`` pragma:
 
 {.emit: """
+#include <stdio.h>
 static int cvariable = 420;
 
 """.}

--- a/tests/niminaction/Chapter8/sdl/sdl_test.nim
+++ b/tests/niminaction/Chapter8/sdl/sdl_test.nim
@@ -18,7 +18,7 @@ renderer.setDrawColor 29, 64, 153, 255
 renderer.clear
 renderer.setDrawColor 255, 255, 255, 255
 
-when defined(c):
+when false: # no long work with gcc 14!
   # just to ensure code from NimInAction still works, but
   # the `else` branch would work as well in C mode
   var points = [

